yetanotherco/aligned_layer
# feat: disable verifiers

**Changes**
This pr introduces proving systems or verifiers blacklisting through a new bitmap field in the `ServiceManager` contract. The batcher and operator, before doing a verification, they make sure that the proving system is indeed whitelisted by calling `isVerifierValid` function in the `ServiceManager` contract. 

It also includes the verification for disabled verifiers in the operators and the batcher logic. For the batcher it will now check for every new block if the disabled verifiers have changed, in the afirmative case it checks all the proofs inside the batcher queue and remove all of them indistinctly if it was a proof using the disabled verifier or not.

**Test**
You can test this following these steps:
- Set up and run all the different components (anvil, aggregator, operator and batcher)
- In another terminal run `make batcher_send_burst_groth16` just to keep generating new proofs and avoid problems trying to submit repeated batches.
- To update the blacklisted verifiers bitmap you can use the following makefile target that sends a transaction to the service manager with the function signature `disableVerifier(uint8)`. The `VERIFIER_ID=3` variable will disable the SP1 verifier:
```
make verifier_disable_devnet VERIFIER_ID=3
```
- Try to send a sp1 proof that it should be rejected by the batcher: `make batcher_send_sp1_task`.
- Whitelist the verifier doing:
```
make verifier_enable_devnet VERIFIER_ID=3
```
- Try to send a sp1 proof and it should be working again: `make batcher_send_sp1_task`.

If you want to try and disable all the other verifiers run the same command mentioned above changing the `VERIFIER_ID`
that is the ID corresponding to the `ProvingSystemId` enum in the `proving_systems.go` file, being `GnarkPlonkBls12_381` the id 0.

**Disabled verifiers in the batch queue testing**

1. Start the Aligned components (anvil, aggregator, and operator).
2. Open the batcher's configuration file at `config-files/config-batcher.yaml` and adjust the following parameters:
   ```
   block_interval: 100000
   batch_size_interval: 4
   ```
   > This ensures the batch won't be submitted while sending valid proofs, but the batch remains open, allowing you to disable the verifier later.
   
3. Run the batcher:  
   `make batcher_start_local`
   
4. Send 3 proofs using any of the verifiers:  
   - `make batcher_send_sp1_task`
   - `make batcher_send_risc0_task`
   - `make batcher_send_groth16_bn254_task`
   
5. Disable the SP1 verifier:  
   `make verifier_disable_devnet VERIFIER_ID=3`
   
6. Verify that all the clients stop and encounter an `BatchReset` error. The batcher's queue should now be empty.

---
This closes #1082
